File: app/src/main/java/deltazero/amarok/filehider/NoMediaFileHider.java
package deltazero.amarok.filehider;

import android.content.Context;
import android.util.Log;

import java.nio.file.Paths;
import java.util.Set;

import deltazero.amarok.utils.MediaStoreHelper;

public class NoMediaFileHider extends BaseFileHider {

    private static final String TAG = "NoMediaFileHider";

    public NoMediaFileHider(Context context) {
        super(context);
    }

    @Override
    protected void process(Set<String> targetDirs, ProcessMethod method) throws InterruptedException {
        for (var dir : targetDirs) {

            if (Thread.interrupted())
                throw new InterruptedException();

            Log.i(TAG, String.format("Processing: %s", dir));

            try {

                var path = Paths.get(dir);
                var nomediaFile = path.resolve(".nomedia").toFile();

                if (method == ProcessMethod.HIDE) {

                    var result = nomediaFile.createNewFile();
                    if (!result) Log.w(TAG, String.format(".nomedia already exist: %s", path));

                } else if (method == ProcessMethod.UNHIDE) {

                    if (!nomediaFile.isFile()) {
                        Log.w(TAG, String.format(".nomedia not file: %s", path));
                        continue;
                    }

                    var result = nomediaFile.delete();
                    if (!result) Log.w(TAG, String.format("Failed to remove .nomedia: %s", path));
                }

            } catch (Exception e) {
                Log.w(TAG, String.format("Error while processing %s: ", dir), e);
            }
        }

        MediaStoreHelper.scan(context, targetDirs);
    }

    @Override
    public void tryToActive(ActivationCallbackListener activationCallbackListener) {
        activationCallbackListener.onActivateCallback(this.getClass(), true, 0);
    }

    @Override
    public String getName() {
        return "NoMedia";
    }
}

File: app/src/main/java/deltazero/amarok/filehider/NoMediaFileHider.kt
package deltazero.amarok.filehider

import android.content.Context
import android.util.Log
import deltazero.amarok.core.ActivationResult
import deltazero.amarok.core.HideAction
import deltazero.amarok.utils.MediaStoreHelper
import java.nio.file.Paths
import kotlin.coroutines.coroutineContext
import kotlinx.coroutines.ensureActive

class NoMediaFileHider(private val context: Context) : FileHider {
  override val mode = FileHiderMode.NOMEDIA
  override val name = "NoMedia"

  override suspend fun activate() = ActivationResult(success = true, msgResId = 0)

  override suspend fun process(targetDirs: Set<String>, action: HideAction) {
    val hide = action is HideAction.Hide
    for (dir in targetDirs) {
      coroutineContext.ensureActive()
      Log.i(TAG, "Processing: $dir")
      try {
        val path = Paths.get(dir)
        val nomediaFile = path.resolve(".nomedia").toFile()
        if (hide) {
          val result = nomediaFile.createNewFile()
          if (!result) Log.w(TAG, ".nomedia already exist: $path")
        } else {
          if (!nomediaFile.isFile) {
            Log.w(TAG, ".nomedia not file: $path")
            continue
          }
          val result = nomediaFile.delete()
          if (!result) Log.w(TAG, "Failed to remove .nomedia: $path")
        }
      } catch (e: Exception) {
        Log.w(TAG, "Error while processing $dir: ", e)
      }
    }
    MediaStoreHelper.scan(context, targetDirs)
  }

  companion object {
    private const val TAG = "NoMediaFileHider"
  }
}
